spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/CarryoverRemoveIterator.java

```diff
@@ -0,0 +1,155 @@
+
+/**
+ * An iterator that removes the carry-over rows from changelog tables within a single Spark task. It
+ * assumes that rows are partitioned by identifier(or all) columns, and it is sorted by both
+ * identifier(or all) columns and change type.
+ *
+ * <p>Carry-over rows are the result of a removal and insertion of the same row within an operation
+ * because of the copy-on-write mechanism. For example, given a file which contains row1 (id=1,
+ * data='a') and row2 (id=2, data='b'). A copy-on-write delete of row2 would require erasing this
+ * file and preserving row1 in a new file. The change-log table would report this as follows,
+ * despite it not being an actual change to the table.
+ *
+ * <ul>
+ *   <li>(id=1, data='a', op='DELETE')
+ *   <li>(id=1, data='a', op='INSERT')
+ *   <li>(id=2, data='b', op='DELETE')
+ * </ul>
+ *
+ * The iterator finds the carry-over rows and removes them from the result. For example, the above
+ * rows will be converted to:
+ *
+ * <ul>
+ *   <li>(id=2, data='b', op='DELETE')
+ * </ul>
+ */
+class CarryoverRemoveIterator extends ChangelogIterator {
+  private final Iterator<Row> rowIterator;
+  private final int[] indicesForIdentifySameRow;
+
+  private Row deletedRow = null;
+  private long deletedRowCount = 0;
+  private Row nextCachedRow = null;
+
+  CarryoverRemoveIterator(Iterator<Row> rowIterator, StructType rowType) {
+    super(rowIterator, rowType);
+    this.rowIterator = rowIterator;
+    this.indicesForIdentifySameRow = generateIndicesForIdentifySameRow(rowType.size());
+  }
+
+  @Override
+  public boolean hasNext() {
+    if (hasDeleteRow() || nextCachedRow != null) {
+      return true;
+    }
+    return rowIterator.hasNext();
+  }
+
+  @Override
+  public Row next() {
+    if (popupDeleteRow()) {
+      deletedRowCount--;
+      return deletedRow;
+    }
+
+    Row currentRow = currentRow();
+
+    if (currentRow.getString(changeTypeIndex()).equals(DELETE) && rowIterator.hasNext()) {
+      // cache the delete row if there is 0 delete row cached
+      if (!hasDeleteRow()) {
+        deletedRow = currentRow;
+        deletedRowCount++;
+      }
+
+      Row nextRow = rowIterator.next();
+
+      if (isSameRecord(currentRow, nextRow)) {
+        if (nextRow.getString(changeTypeIndex()).equals(INSERT)) {
+          deletedRowCount--;
+        } else {
+          deletedRowCount++;
+        }
+      } else {
+        // mark the boundary since the next row is not the same record as the current row
+        nextCachedRow = nextRow;
+      }
+
+      currentRow = null;
+    }
+
+    return currentRow;
+  }
+
+  /**
+   * Pop up the delete rows if there are delete rows cached and the next row is not the same record
+   * or there is no next row.
+   */
+  private boolean popupDeleteRow() {
+    return hitBoundary() && hasDeleteRow();
+  }
+
+  private boolean hitBoundary() {
+    return !rowIterator.hasNext() || nextCachedRow != null;
+  }
+
+  private boolean hasDeleteRow() {
```

Comment:
question: If I understand correctly, this checks if there are any buffered delete rows. If so, can we reflect that in the name?